src/disk_io_thread.cpp

```diff
@@ -1571,57 +1584,66 @@ namespace libtorrent
 	{
 		TORRENT_ASSERT(j->action == disk_io_job::read);
 
-		if (m_settings.get_bool(settings_pack::use_read_cache)
-			&& m_settings.get_int(settings_pack::cache_size) != 0)
+		int ret = m_disk_cache.try_read(j);
+		if (ret >= 0)
 		{
-			int ret = m_disk_cache.try_read(j);
-			if (ret >= 0)
-			{
-				m_stats_counters.inc_stats_counter(counters::num_blocks_cache_hits);
-				DLOG("do_read: cache hit\n");
-				j->flags |= disk_io_job::cache_hit;
-				j->ret = ret;
-				return 0;
-			}
-			else if (ret == -2)
-			{
-				j->error.ec = error::no_memory;
-				j->error.operation = storage_error::alloc_cache_piece;
-				j->ret = disk_io_job::operation_failed;
-				return 0;
-			}
+			m_stats_counters.inc_stats_counter(counters::num_blocks_cache_hits);
+			DLOG("do_read: cache hit\n");
+			j->flags |= disk_io_job::cache_hit;
+			j->ret = ret;
+			return 0;
+		}
+		else if (ret == -2)
+		{
+			j->error.ec = error::no_memory;
+			j->error.operation = storage_error::alloc_cache_piece;
+			j->ret = disk_io_job::operation_failed;
+			return 0;
+		}
 
-			if (check_fence && j->storage->is_blocked(j))
-			{
-				// this means the job was queued up inside storage
-				m_stats_counters.inc_stats_counter(counters::blocked_disk_jobs);
-				DLOG("blocked job: %s (torrent: %d total: %d)\n"
-					, job_action_name[j->action], j->storage ? j->storage->num_blocked() : 0
-					, int(m_stats_counters[counters::blocked_disk_jobs]));
-				return 2;
-			}
+		if (check_fence && j->storage->is_blocked(j))
+		{
+			// this means the job was queued up inside storage
+			m_stats_counters.inc_stats_counter(counters::blocked_disk_jobs);
+			DLOG("blocked job: %s (torrent: %d total: %d)\n"
+				, job_action_name[j->action], j->storage ? j->storage->num_blocked() : 0
+				, int(m_stats_counters[counters::blocked_disk_jobs]));
+			return 2;
+		}
 
-			cached_piece_entry* pe = m_disk_cache.allocate_piece(j, cached_piece_entry::read_lru1);
-			if (pe == NULL)
-			{
-				j->ret = -1;
-				j->error.ec = error::no_memory;
-				j->error.operation = storage_error::read;
-				return 0;
-			}
-			j->flags |= disk_io_job::use_disk_cache;
-			if (pe->outstanding_read)
-			{
-				TORRENT_PIECE_ASSERT(j->piece == pe->piece, pe);
-				pe->read_jobs.push_back(j);
-				return 2;
-			}
+		if (!m_settings.get_bool(settings_pack::use_read_cache)
+			|| m_settings.get_int(settings_pack::cache_size) == 0)
+		{
+			// if the read cache is disabled then we can skip going through the cache
+			// but only if there is no existing piece entry. Otherwise there may be a
+			// partial hit on one-or-more dirty buffers so we must use the cache
+			// to avoid reading bogus data from storage
+			if (m_disk_cache.find_piece(j) == NULL)
```

Comment:
if we *do* find a piece here, we don't return and fall through. then we get to m_disk_cache.allocate_piece() and create a new piece for this job. should the case where we find a piece end up setting pe to point to this piece instead?